br/pkg/utils/register.go

```diff
@@ -0,0 +1,280 @@
+type TaskRegister struct {
+	client    *clientv3.Client
+	ttl       time.Duration
+	secondTTL int64
+	key       string
+
+	// leaseID used to revoke the lease
+	curLeaseID clientv3.LeaseID
+	wg         sync.WaitGroup
+	cancel     context.CancelFunc
+}
+
+// NewTaskRegisterWithTTL build a TaskRegister with key format {RegisterTaskPrefix}/{RegisterTaskType}/{taskName}
+func NewTaskRegisterWithTTL(client *clientv3.Client, ttl time.Duration, tp RegisterTaskType, taskName string) *TaskRegister {
+	return &TaskRegister{
+		client:    client,
+		ttl:       ttl,
+		secondTTL: int64(ttl / time.Second),
+		key:       path.Join(RegisterImportTaskPrefix, tp.String(), taskName),
+
+		curLeaseID: clientv3.NoLease,
+	}
+}
+
+// NewTaskRegister build a TaskRegister with key format {RegisterTaskPrefix}/{RegisterTaskType}/{taskName}
+func NewTaskRegister(client *clientv3.Client, tp RegisterTaskType, taskName string) *TaskRegister {
+	return NewTaskRegisterWithTTL(client, defaultTaskRegisterTTL, tp, taskName)
+}
+
+// Close closes the background task of taskRegister
+func (tr *TaskRegister) Close(ctx context.Context) (err error) {
+	tr.cancel()
+	tr.wg.Wait()
+	if tr.curLeaseID != clientv3.NoLease {
+		_, err = tr.client.Lease.Revoke(ctx, tr.curLeaseID)
+		if err != nil {
+			log.Warn("failed to revoke the lease", zap.Error(err), zap.Int64("lease-id", int64(tr.curLeaseID)))
+		}
+	}
+	return err
+}
+
+func (tr *TaskRegister) grant(ctx context.Context) (*clientv3.LeaseGrantResponse, error) {
+	lease, err := tr.client.Lease.Grant(ctx, tr.secondTTL)
+	if err != nil {
+		return nil, err
+	}
+	if len(lease.Error) > 0 {
+		return nil, errors.New(lease.Error)
+	}
+	return lease, nil
+}
+
+// RegisterTask firstly put its key to PD with a lease,
+// and start to keepalive the lease in the background.
+func (tr *TaskRegister) RegisterTask(c context.Context) error {
+	cctx, cancel := context.WithCancel(c)
+	tr.cancel = cancel
+	lease, err := tr.grant(cctx)
+	if err != nil {
+		return errors.Annotatef(err, "failed grant a lease")
+	}
+	tr.curLeaseID = lease.ID
+	_, err = tr.client.KV.Put(cctx, tr.key, "", clientv3.WithLease(lease.ID))
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// KeepAlive interval equals to ttl/3
+	respCh, err := tr.client.Lease.KeepAlive(cctx, lease.ID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	tr.wg.Add(1)
+	go tr.keepaliveLoop(cctx, respCh)
+	return nil
+}
+
+func (tr *TaskRegister) keepaliveLoop(ctx context.Context, ch <-chan *clientv3.LeaseKeepAliveResponse) {
+	defer tr.wg.Done()
+	const minTimeLeftThreshold time.Duration = 20 * time.Second
+	var timeLeftThreshold time.Duration = tr.ttl / 4
+	if timeLeftThreshold < minTimeLeftThreshold {
+		timeLeftThreshold = minTimeLeftThreshold
+	}
+	failpoint.Inject("brie-task-register-always-grant", func(_ failpoint.Value) {
+		timeLeftThreshold = tr.ttl
+	})
+	var lastUpdateTime time.Time = time.Now()
+	var err error
```

Comment:
Can we merge these variable definitions into one block like:
```go
var (
  foo = somewhat
  bar = somewhat
) 
```

Reply:
done